mcf-browser/src/TimelineExplorer.tsx

```diff
@@ -0,0 +1,390 @@
+/**
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, {Component} from 'react';
+import Select, {MultiValue} from 'react-select';
+
+import {Series} from './back-end/time-series';
+import {Grouping, groupSeriesByLocations, renderTimeGraph} from './utils';
+import {LoadingSpinner} from './LoadingSpinner';
+import {getName} from './back-end/utils';
+import {PageBar} from './PageBar';
+
+const STAT_VARS_PER_PAGE = 10;
+const INITIAL_NUM_LOCATIONS = 5;
+
+interface LocationMapping {
+  [dcid: string]: string;
+}
+
+interface SelectOption {
+  value: string;
+  label: string;
+}
+
+interface TimelineExplorerPropType {
+  /**
+  * Passes the data to be plotted
+  */
+  data: Series[];
+
+  /**
+  * Indicates if uploaded files are currently being parsed.
+  */
+  loading: boolean;
+}
+
+interface TimelineExplorerStateType {
+  /** The currently selected locations to filter by */
+  locations: string[];
+
+  /** A list of possible locations to select from */
+  locationOptions: SelectOption[]
+
+  /** The key for the select component to tell it
+  * to re-render when the locations are loaded in
+  */
+  selectKey: string;
+
+  /** A mapping from a location dcid to its name */
+  locationMapping: LocationMapping;
+
+  /** The list of groups to plot, grouped by stat var */
+  statVarGroups: Series[][];
+
+  /** Tracks which page the user is currently viewing (0-indexed) */
+  page: number;
+}
+
+/** Component to display the timeline explorer */
+class TimelineExplorer extends Component<
+  TimelineExplorerPropType,
+  TimelineExplorerStateType
+> {
+  /** Constructor for class, sets initial state
+    * @param {TimelineExplorerPropType} props the props passed in by parent
+    * component
+    */
+  constructor(props: TimelineExplorerPropType) {
+    super(props);
+    this.state = {
+      locations: [],
+      locationOptions: [],
+      selectKey: 'unloaded',
+      locationMapping: {},
+      statVarGroups: [],
+      page: 0,
+    };
+  }
+
+  /**
+   * Update the state variables whenever necessary given
+   * the current value of this.props
+   */
+  updateState() {
+    const locationsPromise = this.getAllLocations();
+    locationsPromise.catch(
+        () => {
+          console.log(
```

Comment:
usually console.log is for logging things for a developer. If this is an error you want to show the user in the UI, then maybe use `alert()`

Reply:
Changed!